src/mcs/montecarlo.jl

```diff
@@ -363,57 +400,58 @@ If provided, `scenario_args` must be a `Vector{Pair}`, where each `Pair` is a sy
 `Vector` of arbitrary values that will be meaningful to `scenario_func`, which must have
 the signature:
 
-    scenario_func(sim::Simulation, tup::Tuple)
+    scenario_func(sim_inst::SimulationInstance, tup::Tuple)
 
 By default, the scenario loop encloses the simulation loop, but the scenario loop can be
 placed inside the simulation loop by specifying `scenario_placement=INNER`. When `INNER` 
 is specified, the `scenario_func` is called after any `pre_trial_func` but before the model
 is run.
+
+Returns the type `SimulationInstance` that contains a copy of the original `SimulationDef`,
+along with mutated information about trials, in addition to the model list and 
+results information.
 """
-function run_sim(sim::Simulation{T}; 
-                 trials::Union{Nothing, Int, Vector{Int}, AbstractRange{Int}}=nothing,
-                 models_to_run::Int=length(sim.models),     # run all models by default
+function Base.run(sim_def::SimulationDef{T}, models::Union{Vector{Model}, Model}, samplesize::Int;
                  ntimesteps::Int=typemax(Int), 
-                 output_dir::Union{Nothing, AbstractString}=nothing, 
+                 trials_output_filename::Union{Nothing, AbstractString}=nothing, 
+                 results_output_dir::Union{Nothing, AbstractString}=nothing, 
                  pre_trial_func::Union{Nothing, Function}=nothing, 
                  post_trial_func::Union{Nothing, Function}=nothing,
                  scenario_func::Union{Nothing, Function}=nothing,
                  scenario_placement::ScenarioLoopPlacement=OUTER,
-                 scenario_args=nothing) where T <: AbstractSimulationData
+                 scenario_args=nothing,
+                 results_in_memory::Bool=true) where T <: AbstractSimulationData
+            
+    # Quick check for results saving
+    if (!results_in_memory) && (results_output_dir===nothing)
+        error("The results_in_memory keyword arg is set to ($results_in_memory) and 
+        results_output_dir keyword arg is set to ($results_output_dir), thus 
+        results will not be saved eitehr in memory or in a file.")
+    end
+
+    # Initiate the SimulationInstance and set the models and trials for the copied 
+    # sim held within sim_inst
+    sim_inst = SimulationInstance{typeof(sim_def.data)}(sim_def)
+    set_models!(sim_inst, models)
```

Comment:
If you define `SimulationInstance` in terms of a type `T`, you can avoid this explicit call to `typeof(sim_def.data)`. For example:
```
julia> struct SimDef{T}
         data::T
         end

julia> struct SimInst{T}
         sim_def::SimDef{T}
         end

julia> sd = SimDef((10, 20))
SimDef{Tuple{Int64,Int64}}((10, 20))

julia> SimInst(sd)
SimInst{Tuple{Int64,Int64}}(SimDef{Tuple{Int64,Int64}}((10, 20)))
```

Reply:
So given I already have this:

Reply:
Yes, and I think you can drop the where clause from
```
sim_def::SimulationDef{T} where T <: AbstractSimulationData
```
since the constructor enforces this.

Reply:
Ah ok, I tried this and got some complaints about my `SimulationInstance` call to `new` but I'll play with it a bit more, the elusive `{T}` syntax ...

Reply:
You might need `new{T}(args...)`